Source/Orts.Simulation/Simulation/AIs/AI.cs

```diff
@@ -537,6 +537,8 @@ private void PrerunAI(int playerTrainOriginalTrain, TTTrain.FormCommand playerTr
                         Simulator.Signals.Update(true);
                         clockTime = runTime;
                         runTime += deltaTime;
+                        if (cancellation.IsCancellationRequested) // ping loader watchdog
```

Comment:
The first part is a for loop, as it is known when it should terminate (at the start time of the player train).
The second part (this loop) however is 'open ended'. The train can not start, it is delayed, but for how much is not clear. Setting up a 'for' loop would be misleading, as that creates the impression it is to run upto a certain point - but it is not. The check on midnight is an 'emergency exit' and terminates the program immediately.
Actually this is not quite correct for a train may be started after midnight and if this train is running late, the program would abort and not allow it to start.
This is on my list to review but I have not yet puzzled out what the proper 'emergency exit' criterium should be.

Reply:
I disagree it would be misleading; instead, it would clearly indicate this loop terminates. The problem with non-terminating loops like `while` is that now you've added a loader ping, if at any point in the future the code inside the loop changes and it - intentionally or, more likely, accidentally - becomes possible for it to not terminate, you've broken the ability for Open Rails to detect that and report a bug.

This one I'll let go but please remember this for future code.